### DISCUSSION TITLE
Routes registered via `defineRoutes()` not working

### DISCUSSION THREAD

**Discussion by clink-aaron** [orchestral/testbench-dusk#64] [Ideas]
- Testbench Version: 6.2.4
- Laravel Version: 8.75
- PHP Version: 8.0.12
- Database Driver & Version:
N/A
### Description:
I've found that routes defined in tests via `defineRoutes()` are not reachable via testbench-dusk`. Oddly, they _are_ reachable via the standard test suite. 

I would very much like to avoid defining routes in a package I'm developing that exists just for the sake of testing. Being able to define routes that only exist for testing is a must. 

### Steps To Reproduce:
I've uploaded a barebones repo here: https://github.com/clink-aaron/testbench-demo that clearly demonstrates the issue. I've copied the instructions into the readme there as well.

First install the dependencies:

     make setup # this runs docker-compose run --rm testbench composer install

Run the tests

     make test # this runs docker-compose run --rm --service-ports --use-aliases phpunit

- Observe that the test `BasicTest::testRoute()` passes
- Observe that the test `BrowserTest::testRoute()` passes
- Observe that the test `BrowserTest::testBrowser()` fails
- Observe that the `tests/Browser/sreenshots/` directory contains an image indicating a 404 error

I would expect that routes registered would be available. I've not yet found a workaround for this issue.

**Comment by crynobone** [comment]
`defineRoutes()` is not designed for testbench-dusk as tests and web service instance is not executed from the same process. 

However, feel free to implement a PR to change this if you can.

**Comment by Keoghan** [comment]
I think a work around would be to define your routes in a TestProvider that is booted only in the test environment, by only registering with Testbench-Dusk

You might also be able to use the tweakApplication method to inject the routes, since that is used when booting the server using the test class.

**Comment by clink-aaron** [comment]
@Keoghan that does indeed work, but with a catch. 

I was hoping it would be as simple as this:
```
  protected function setUp(): void
  {
      parent::setUp();
      $this->tweakApplication(function ($app, $config) {
          $this->defineRoutes($app['router']);
      });
  }

  protected function defineRoutes($router)
  {
      $router->get('/hi', fn() => 'hello');
  }
```
However, upon further investigation, I found that the way the server app stuff registered in the `tweakApplication` is actually serializing the whole thing to a file. the entire context of `$this` was being serialized along with it. I played with bumping memory limits, but after 1 gig I think it's just not a viable solution. 

What _does_ work is just defining the routes directly in the method:
```
  protected function setUp(): void
  {
      parent::setUp();
      $this->tweakApplication(function ($app, $config) {
          $app['router']->get('/hi', fn() => 'hello');
      });
  }
```

So, as much as I'd like to submit a PR to provide this functionality, I'm not sure of a way to deal with the serialization. This is a shame, because it would be a really simple and universal way to allow routes to be registered regardless of how tests are being run.

**Comment by clink-aaron** [comment]
UPDATE: I'm dumb. I can just call it statically. this works just fine:
```
protected function setUp(): void
{
    parent::setUp();
    $this->tweakApplication(function ($app, $config) {
        self::defineRoutes($app['router']);
    });
}

protected function defineRoutes($router)
{
    $router->get('/hi', fn() => 'hello');
}
```
as far as a PR goes, just dropping that in the base `setUp()` method should work great yeah?
